src/pieces.rs

```diff
@@ -89,24 +54,18 @@ impl<'a> Tower<'a> {
     ///    For example, (Your) Pawn, (Your) Gold, (Your) Gold is disallowed
     ///    but (Your) Pawn, (Your) Gold, (Enemy) Gold is fine
     pub fn is_valid(&self) -> bool {
-        match (self.bottom, self.mid, self.top) {
+        use pieces::Tower::*;
+        match *self {
             // Empty towers are obviously fine!
-            (None, None, None) => true,
-            // Cases where the tower clearly isn't a tower (ie: bottom pieces missing)
-            (None, _, _) => false,
-            (Some(_), None, Some(_)) => false,
+            Empty => true,
             // Towers of just one piece can never have two pieces of the same type
-            (Some(_), None, None) => true,
-            (Some(bottom), Some(middle), None) => {
+            Single(_) => true,
             // Towers of two mustn't have the pieces be the same type and from same player
-                return !same_type_and_player(bottom, middle)
-            }
+            Double(bottom, middle) => !same_type_and_player(bottom, middle),
```

Comment:
`==` will compare every field in the struct, but the only thing that we want to compare are the pointers in `belongs_to` and the side that is currently active. 

From https://mmmmalo.tumblr.com/post/74510568781/rules-of-gungi:
> Within a tower, two pieces of the same type and team cannot coexist. It is not against the rules to occupy the tower of an enemy that contains a piece of the same type. Here are some examples:
>•Pawn-Pawn-Gold is prohibited:
>•Pawn-Pawn-Pawn is prohibited:
>•Pawn-Bronze-Gold is allowed:
>•(Your) Pawn – (Your) Gold – (Enemy) Gold is allowed:

(Side note: Pawn has different variations. Pawn is the front side but the back side maybe be Bronze, Silver or Gold. A complete list is in the rules) Thus what matters is the "current type" (this is why `current_type` exists).

Perhaps I should implement a custom `Eq` and `PartialEq`?